transport/transporters/kafka/transporter/transporter_test.go

```diff
@@ -0,0 +1,407 @@
+	}
+	stats.VerifyStats(t, statsChan, expected)
+
+	time.Sleep(50 * time.Millisecond)
+
+}
+
+func TestInputClosed(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	defer resetTimeSource()
+
+	in := make(chan transport.Batch, 1000)
+	txns := make(chan *ordered_map.OrderedMap, 1000)
+	statsChan := make(chan stats.Stat, 1000)
+	mockProducer := mockProducer(t)
+	mockTime := utils_mocks.NewMockTimeSource(mockCtrl)
+	TimeSource = mockTime
+	sh := shutdown.NewShutdownHandler()
+	topic := "test"
+
+	tp := NewTransporter(sh, in, statsChan, txns, *log, mockProducer, topic)
+
+	go tp.StartTransporting()
+
+	// Wait for transporter to start
+	time.Sleep(time.Millisecond * 5)
+
+	// Close input
+	close(in)
+	time.Sleep(time.Millisecond * 5)
+
+	// Verify shutdown
+	_, ok := <-txns
+	fmt.Println("checking if channel is closed")
+	if ok {
+		assert.Fail(t, "output channel not properly closed")
+	}
+
+	_, ok = <-sh.TerminateCtx.Done()
+	if ok {
+		assert.Fail(t, "context not cancelled")
+	}
+}
+
+func TestTerminationContext(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	defer resetTimeSource()
+
+	in := make(chan transport.Batch, 1000)
+	txns := make(chan *ordered_map.OrderedMap, 1000)
+	statsChan := make(chan stats.Stat, 1000)
+	mockProducer := mockProducer(t)
+	mockTime := utils_mocks.NewMockTimeSource(mockCtrl)
+	TimeSource = mockTime
+	sh := shutdown.NewShutdownHandler()
+	topic := "test"
+
+	tp := NewTransporter(sh, in, statsChan, txns, *log, mockProducer, topic)
+
+	go tp.StartTransporting()
+
+	// Wait for transporter to start
+	time.Sleep(time.Millisecond * 5)
+
+	// Cancel
+	sh.CancelFunc()
+	time.Sleep(time.Millisecond * 5)
+
+	// Verify shutdown
+	_, ok := <-txns
+	if ok {
+		assert.Fail(t, "output channel not properly closed")
+	}
+}
+
+func TestPanicHandling(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	defer resetTimeSource()
+
+	in := make(chan transport.Batch, 1000)
+	txns := make(chan *ordered_map.OrderedMap, 1000)
+	statsChan := make(chan stats.Stat, 1000)
+	mockProducer := mockProducer(t)
+	mockTime := utils_mocks.NewMockTimeSource(mockCtrl)
+	TimeSource = mockTime
+	sh := shutdown.NewShutdownHandler()
+	topic := "test"
+	batchSize := 1
+	maxMessageBytes := 1000000
+
+	tp := NewTransporter(sh, in, statsChan, txns, *log, mockProducer, topic)
+	b := batch.NewKafkaBatch(topic, "", batchSize, maxMessageBytes)
+
+	marshalledMessage := marshaller.MarshalledMessage{
+		Operation:    "INSERT",
+		Json:         []byte("data"),
+		TimeBasedKey: "123",
+		WalStart:     1234,
+		Transaction:  "123",
+	}
+	_, _ = b.Add(&marshalledMessage)
+
+	mockProducer.ExpectSendMessageWithMessageCheckerFunctionAndSucceed(testMessageCheckerWithPanic())
+
+	mockTime.EXPECT().UnixNano().Return(int64(0))
+
+	// Start Test
+	in <- b
+
+	go tp.StartTransporting()
+
+	// Sleep a little to ensure retries had time to run
+	time.Sleep(time.Millisecond * 25)
```

Comment:
I realize much of the code has sleeps but I encourage using channels with wait timeouts instead of sleeps so you avoid timing/race condition issues in the tests. You can do this via `DoAndReturn` eg.

Reply:
ahh good point. I've used either waiting on the transaction channel or waiting for the transporter to shutdown to progress through the unit test, as opposed to sleeping. While waiting for shutdown, I fail the test after a certain timeout period.